src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -102,14 +115,60 @@ try
             _r.deviceContext->OMSetRenderTargets(1, _r.renderTargetView.addressof(), nullptr);
             _r.deviceContext->Draw(3, 0);
         }
+    }
 
-        // > IDXGISwapChain::Present: Partial Presentation (using a dirty rects or scroll) is not supported
-        // > for SwapChains created with DXGI_SWAP_EFFECT_DISCARD or DXGI_SWAP_EFFECT_FLIP_DISCARD.
-        // ---> No need to call IDXGISwapChain1::Present1.
+    // See documentation for IDXGISwapChain2::GetFrameLatencyWaitableObject method:
+    // > For every frame it renders, the app should wait on this handle before starting any rendering operations.
+    // > Note that this requirement includes the first frame the app renders with the swap chain.
+    assert(debugGeneralPerformance || _r.frameLatencyWaitableObjectUsed);
+
+    if (_r.dirtyRect != fullRect)
+    {
+        auto dirtyRectInPx = _r.dirtyRect;
+        dirtyRectInPx.left *= _r.fontMetrics.cellSize.x;
+        dirtyRectInPx.top *= _r.fontMetrics.cellSize.y;
+        dirtyRectInPx.right *= _r.fontMetrics.cellSize.x;
+        dirtyRectInPx.bottom *= _r.fontMetrics.cellSize.y;
+
+        RECT scrollRect{};
+        POINT scrollOffset{};
+        DXGI_PRESENT_PARAMETERS params{
+            .DirtyRectsCount = 1,
+            .pDirtyRects = dirtyRectInPx.as_win32_rect(),
+        };
+
+        if (_r.scrollOffset)
+        {
+            scrollRect = {
+                0,
+                std::max<til::CoordType>(0, _r.scrollOffset),
+                _r.cellCount.x,
+                _r.cellCount.y + std::min<til::CoordType>(0, _r.scrollOffset),
+            };
+            scrollOffset = {
+                0,
+                _r.scrollOffset,
+            };
+
+            scrollRect.top *= _r.fontMetrics.cellSize.y;
+            scrollRect.right *= _r.fontMetrics.cellSize.x;
+            scrollRect.bottom *= _r.fontMetrics.cellSize.y;
+
+            scrollOffset.y *= _r.fontMetrics.cellSize.y;
+
+            params.pScrollRect = &scrollRect;
+            params.pScrollOffset = &scrollOffset;
+        }
+
+        THROW_IF_FAILED(_r.swapChain->Present1(1, 0, &params));
+    }
+    else
+    {

```

Comment:
5. Flip-Discard and `Present()` is less energy efficient than
Flip-Sequential and `Present1()`

Reply:
(Presumably, this is for _that thread with the composition team_)